proposals/schism-eval-fib/README.md

```diff
@@ -0,0 +1,204 @@
+# Schism eval('fib(25)') benchmark
+
+## Overview
+
+This benchmark runs a Scheme interpreter (`eval`), compiled by the
+[Schism](https://github.com/google/schism) self-hosted Scheme
+implementation.  The interpreter is used to compute and print the 25th
+fibonacci number.
+
+This branch of Schism represents Scheme values using `anyref` values
+originating in the embedder.  The embedder has to supply allocator, type
+predicate, and field access instructions for the different Scheme value
+types supported by Schism: small integers, characters, strings, symbols,
+pairs, and closures.  Schism provides a run-time written in JavaScript
+which uses tagged JavaScript numbers for Scheme integers and characters,
+tagged JavaScript strings for Scheme strings and symbols, and JavaScript
+objects for pairs and closures.
+
+This benchmark largely tests the performance of `anyref`, host-side
+allocation and GC, and JavaScript strings.  It also stress-tests calls
+between WebAssembly and JS, as well as inter-WebAssembly calls.
+
+## Possible benchmark evolution
+
+The benchmark could be made more suitable in a few ways:
+
+ * Eq-able reference types would be useful.  It would be nice to be able
+   to compare anyref values by identity, though what that means for host
+   strings is a question.
+
+ * We could use tail calls.
+
+ * `i31ref` would be really useful.
+
+ * The Schism compiler can improve in many ways.  Notably the current
+   version doesn't support static data, which means that it allocates
+   more than usual at run-time in order to create literals anew each
+   time they appear.  We could take a version of `eval` compiled from a
+   newer version of Schism that doesn't have this problem.
+
+ * We could use GC types to avoid calling out so much to the host.
+   Would need implementation of GC in an engine though.
+
+## Benchmark kind
+
+I think we can consider this benchmark to be a "kernel".
```

Comment:
How big is the module that is generated? Am I right in assuming that the Schism implementation compiles the fib program to a single module, and that that module doesn't include much of the language runtime?

Reply:
Ah, I see the binary is ~20KiB. That's probably kernel territory.

It'd be good for us to think over the definition of what a kernel is. E.g. does it have a particularly small "working set size" in terms of number of functions or bytecodes?

Reply:
Right.  Schism currently compiles by first adding the user code to a set of primitives, and does not do dead-code elimination.  This would matter for other benchmarks but given that `eval.ss` itself adds most of those primitives to the default environment, I don't think DCE would change much here.

I am not certain what percentage of functions will be executed for this particular input (the fibonacci function).  Probably a fair bit just given that `eval` needs to parse and expand the input.

I think the problem is to have rules to differentiate a kernel from a microbenchmark.  Probably before merging anything we should add guidelines about how to categorize a given benchmark.  Actually, what is the purpose of differentiating between kernels and microbenchmarks?